### ISSUE TITLE
Shifting array values right with memoryview clears array

### ISSUE THREAD

**Issue by fragmuffin** [micropython/micropython#6244]
When shifting values in an `array` +1 index (to the right) using memoryview, 0th index value is carried through the array.

**cPython**

```python
>>> from array import array
>>> a = array('I', range(10))
>>> mv = memoryview(a)
>>> a  # before shifting right
array('I', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> mv[1:] = mv[:-1]
>>> a  # after
array('I', [0, 0, 1, 2, 3, 4, 5, 6, 7, 8])  # <-- looks good
```

**micropython**
```python
>>> from array import array

# shift left
>>> a = array('I', range(10))
>>> mv = memoryview(a)
>>> a  # before shifting right
array('I', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> mv[1:] = mv[:-1]
>>> a  # after
array('I', [0, 0, 0, 0, 0, 0, 0, 0, 0, 0])  # <-- 0th index value carried through entire array

# shift right
>>> a = array('I', range(10))
>>> mv = memoryview(a)
>>> a  # before shifting left
array('I', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> mv[:-1] = mv[1:]
>>> a  # after
array('I', [1, 2, 3, 4, 5, 6, 7, 8, 9, 9])  # <-- looks good
```

**Comment by fragmuffin** [comment]
Workaround: populate memoryview from a copy of the array:

```python
>>> mv[1:] = a[:-1]
>>> a
array('I', [0, 0, 1, 2, 3, 4, 5, 6, 7, 8])
```

Or, for a large `array`, manually loop through it backwards:
```python
>>> for i in range(len(a)-1, 0, -1):
...     a[i] = a[i-1]
```

**Comment by peterhinch** [comment]
It also happens with a bytearray:
```python
a = bytearray(range(10))
mv = memoryview(a)
mv[1:] = mv[:-1]
print(a)
```

**Comment by davehylands** [comment]
This might be a case of using memcpy versus memmove in the C code.

**Comment by dpgeorge** [comment]
Thanks @fragmuffin for the report, @peterhinch for the extra test, and @dhylands for the hint about the fix.  Should be fixed by #6249

**Comment by dpgeorge** [comment]
Fixed by a853fff838c1f995b9eeb0ebef5f117cfa3d5f0a
